Allow repeatable interceptor bindings

Ladicek opened the pull request:
Fixes #13332.

Ladicek committed:
Allow repeatable interceptor bindings

manovotn commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:
```diff
@@ -449,9 +449,11 @@ private void putLifecycleInterceptors(Map<InterceptionType, InterceptionInfo> li
-                        && bindings.stream().noneMatch(e -> e.name().equals(a.name())))
-                .forEach(a -> bindings.add(a));
+                .flatMap(a -> beanDeployment.extractInterceptorBindings(a).stream())
+                // TODO this is weird -- either we allow multiple class-level interceptor bindings of the same type,
```
Isn't having multiple bindings of the same type (and all on class level) an error according to specification?

Ladicek replied:
That's how I read the spec, indeed -- but it doesn't look like we enforce it. (At least not yet.)

manovotn replied:
As per Zulip conversation - this should be validated and the spec requires this because you can, for instance, declare the same binding with member on a class and on a stereotype that this class uses leading to a conflict.

mkouba replied:
+1 for validation and potentially a build failure, i.e. repeatable bindings are only allowed on methods which should be documented here https://quarkus.io/guides/cdi-reference#nonstandard_features

Ladicek replied:
OK, only supporting repeatable interceptor bindings on methods, that's an interesting proposition. @manovotn does perhaps Weld do the same? That would explain why there's no test for repeatable interceptor bindings on classes.

mkouba replied:
> does perhaps Weld do the same? That would explain why there's no test for repeatable interceptor bindings on classes.

We should find out ;-)

manovotn replied:
Did you just volunteer to try that you? :-P

manovotn replied:
A quick remake of the existing Weld test shows me that interception happens even if I put repeatable annotation on a class-level of the bean.

Ladicek replied:
Good to know. Doesn't necessarily mean we have to do the same, but I think the current behavior is actually pretty reasonable. (Once you make peace with the fact that whenever you put an interceptor binding on a method, all "inherited" interceptor bindings of the same type are ignored :-) )

manovotn replied:
> Doesn't necessarily mean we have to do the same

No, but for most people familiar with CDI, it would then behave in an "expected manner". E.g. if they used it before, they likely did so on Weld and aligning it would provide consistency. So I'd be +1 for that.

mkouba commented on independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedRepeatableInterceptorBindingTest.java:
```diff
@@ -0,0 +1,148 @@
+    public void classLevelInterceptor() {
+        ClassInterceptedBean bean = Arc.container().instance(ClassInterceptedBean.class).get();
+
+        assertEquals(10, bean.foo());
```
Shouldn't the expected value be `11` because `ClassInterceptedBean#foo()` declares `@MyBinding("foo")` and `@MyBinding("bar")` is inherited from the superclass?

Ladicek replied:
Exactly the kind of feedback I was looking for!

My understanding is that if you declare an interceptor binding on a method, interceptor bindings of the same type coming from the class and all superclasses (and even stereotypes I think) are ignored.

This is per https://jakarta.ee/specifications/interceptors/2.0/interceptors-spec-2.0.html#binding_an_interceptor_to_a_component:

> An interceptor binding declared on a method or constructor replaces an interceptor binding of the same type declared at class level or inherited from a superclass [7].

mkouba replied:
You're right but it's not very intuitive :-(. This interceptors rule is another one that was created before repeatable annotations were introduced...

Ladicek replied:
Yea I totally agree, the interceptors spec needs an update wrt repeatable annotations and default interface methods.

mkouba replied:
Yet another reason to ditch the support of the class-level repeatable bindings...

manovotn replied:
Agree. The test is "correct" per spec but it doesn't follow common sense.
Since the specification is blurred in this regard (e.g. repeatable ann), we may IMO choose to go either way.
FTR I did another quick test in Weld and Weld sides with Ladislav on this one as it also chooses not to inherit repeating annotation if declared on the current class.

mkouba commented on independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/bindings/repeatable/InheritedStereotypeWithRepeatableInterceptorBindingTest.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @MyBinding("foo")
+    @MyBinding("bar")
```
Hm, I don't think this should be supported. A stereotype with interceptor bindings is inherently a class-level-only concept and so I would expect the same limitations as with bindings declared on a class...

Ladicek replied:
The stereotype is class-only.

The test also has methods with the same interceptor binding, verifying that if a method has an interceptor binding, other interceptor bindings (from the class, from superclasses, from stereotype) don't apply, but that's a valid case IMHO?

Ladicek replied:
(Ah and this test is `@Disabled`, because we don't do stereotype inheritance yet. But it's still valid to discuss I believe.)

mkouba replied:
My point is that a stereotype with multiple bindings of the same type should probably result in a definition exception...

Ladicek commented:
I have removed the support for class-level (and stereotype-level) repeatable interceptor bindings, together with the corresponding tests, and added a short section into the documentation.

I didn't add validatation for the "multiple class-level interceptor bindings of the same type" situation. We didn't do that before, so it's not a regression :-) If you think this must go in, let me know and I'll take a look. (Next year, probably :-) )

mkouba commented:
> I have removed the support for class-level (and stereotype-level) repeatable interceptor bindings, together with the corresponding tests, and added a short section into the documentation.

+1

> I didn't add validatation for the "multiple class-level interceptor bindings of the same type" situation. We didn't do that before, so it's not a regression :-) If you think this must go in, let me know and I'll take a look. (Next year, probably :-) )

This should be addressed in a separate issue/PR anyway.

Ladicek commented:
CI passed in my fork, so marked ready for review.

mkouba commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -734,6 +734,29 @@ class SharedService {
+The corresponding interceptor binding is called `@Cached` and is declared as `@Repeatable`.
+In this example, the interceptor will store the method's return value in two caches, because we add `@Cached` twice.
+
+Repeatable interceptor bindings declared on classes and stereotypes are not supported, because there are some open questions around interactions with the Interceptors specification.
```
Maybe we should clarify that repeatable interceptor bindings _declared on interceptor classes_ are supported?

Ladicek replied:
So this can get misleading pretty quickly IMHO. You always need to put an interceptor binding on two places:

- on the interceptor class;
- on the place that should be intercepted.

I don't know if there's a good name for the 2nd item, because that's what we would need here.

I don't think we should say anything like "Repeatable interceptor bindings declared on classes are not supported, unless the class is an interceptor class", because that conflates the two concepts.

mkouba replied:
>     * on the interceptor class;

-> **Declaring Interceptor Bindings of an Interceptor**

>     * on the place that should be intercepted.

-> **Binding an Interceptor to a Component**

This is the spec wording. So I think that we support "Repeatable interceptor bindings declared on methods of component classes and interceptor classes"...

Ladicek replied:
Gotcha, I'll try to figure something out.

Ladicek replied:
WDYT about this?

```adoc
=== Repeatable interceptor bindings

Quarkus has a limited support for `@Repeatable` interceptor binding annotations.

When binding an interceptor to a component, you can declare multiple `@Repeatable` annotations on methods.
Repeatable interceptor bindings declared on classes and stereotypes are not supported, because there are some open questions around interactions with the Interceptors specification.
This might be added in the future.

As an example, suppose we have an interceptor that stores method's return values into a cache.
The corresponding interceptor binding would be called `@Cached` and would be declared as `@Repeatable`.
If we wanted to store the return value into two caches at the same time, we would add `@Cached` twice:

[source,java]
----
@ApplicationScoped
class AllEncompassingService {
  @Cached(cacheName = "foo")
  @Cached(cacheName = "bar")
  BigData heavyComputation() {
    // ...
  }
}
----

This is how interceptors are used.
What about creating an interceptor?

When declaring interceptor bindings of an interceptor, you can add multiple `@Repeatable` annotations to the interceptor class as usual.
This is useless when the annotation members are `@Nonbinding`, as would be the case for the `@Cached` annotation, but is important otherwise.

For example, suppose we have an interceptor that can automatically log method invocations to certain targets.
The interceptor binding annotation `@Logged` would have a member called `target`, which specifies where to store the log.
Our implementation could be restricted to console logging and file logging:

[source,java]
----
@Interceptor
@Logged(target = "console")
@Logged(target = "file")
class NaiveLoggingInterceptor {
  // ...
}
----

Other interceptors could be provided to log method invocations to different targets.
```

gwenneg reviewed (commented):
This looks good, thanks for taking care of it! I added a few comments.

Ladicek commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -734,6 +734,29 @@ class SharedService {
+----
+@ApplicationScoped
+class AllEncompassingService {
+  @Cached(cacheName = "foo") // <1>
```
AsciiDoc actually understands `// <1>` and removes the `//`. Much better, since my IDE knows it's a Java code in that `.adoc` file and gives me syntax highlight and errors and everything -- I don't know why noone else does this! :-)

gwenneg replied:
That's good to know, thanks for the info!

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -409,24 +412,42 @@ ClassInfo getQualifier(DotName name) {
+    /**
+     * Extracts interceptor bindings from given annotation instance.
+     * This returns a collection because in case of repeating interceptor bindings there can be multiple.
+     * For most instances this will be a singleton instance (if given annotatation is an interceptor binding) or
```
Ah, I'll take a look at that, I copied the existing method without much thought :-)

Ladicek replied:
Fixed this.

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -151,6 +153,7 @@
                 }
             }
         }
+        this.repeatingInterceptorBindingAnnotations = findContainerAnnotations(interceptorBindings, this.beanArchiveIndex);
```
Thanks, I'll do that!

Ladicek replied:
I'm looking at it again and I think this is a pretty good name. It takes annotation types and returns the containing annotation types (for each annotation in the original set that is `@Repeatable`). The term "containing annotation type" is what `@Repeatable` javadoc uses, so "container annotation" is fairly close.

I could rename to `findContainingAnnotationTypesForRepeatableAnnotationTypes` or something similarly long, but I don't think that's necessary. I've just renamed the parameters and local variables, as the method was originally written only for qualifiers, but it's now used in another setting as well.

gwenneg replied:
Now that I checked the `Repeatable` javadoc, I agree with you. The current name is good enough and I don't think the extra long and explicit version would improve the code readability. Let's leave it that way! :)

Ladicek commented:
Actually, I really need to move to some other work, so I'll leave this to Monday, if everyone's fine with that :-)

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java:
```diff
@@ -409,24 +412,42 @@ ClassInfo getQualifier(DotName name) {
+     * For most instances this will be a singleton instance (if given annotatation is an interceptor binding) or
+     * an empty list for cases where the annotation is not an interceptor binding.
+     *
+     * @param annotation instance to be inspected
```
It's either "to be inspected", or "to be extracted from". In my (arguably poor) understanding of English, "annotation instance to be extraced" means it's already a result of the extraction process, which it's not; it's an input to the extraction process.

gwenneg replied:
You're right and on second thoughts I don't think there's a problem with this line. I'm fine with the current version.

Ladicek commented:
Rebased and pushed a commit that addresses the documentation feedback and @gwenneg's comments (most of them, anyway).

Ladicek commented:
Rebased and resolved the conflict.

gwenneg commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -734,6 +734,29 @@ class SharedService {
+@ApplicationScoped
+class AllEncompassingService {
+  @Cached(cacheName = "foo") // <1>
+  @Cached(cacheName = "bar")
```
There's another reason why `@Cached` (even the imaginary version) might not be the best pick for that example: it doesn't really make sense to repeat an annotation aimed at retrieving data from a cache because having multiple caches on a single method would raise a lot of questions. That's probably why most (if not all) annotations-based caching API do not allow this.

Ladicek replied:
Yea I know it's stupid. The other example too. If you have a better one, I'm all ears :-)

gwenneg commented on docs/src/main/asciidoc/cdi-reference.adoc:
```diff
@@ -735,6 +735,52 @@ class SharedService {
+Repeatable interceptor bindings declared on classes and stereotypes are not supported, because there are some open questions around interactions with the Interceptors specification.
+This might be added in the future.
+
+As an example, suppose we have an interceptor that stores method's return values into a cache.
+The corresponding interceptor binding would be called `@Cached` and would be declared as `@Repeatable`.
+If we wanted to store the return value into two caches at the same time, we would add `@Cached` twice:
+
+[source,java]
+----
+@ApplicationScoped
+class AllEncompassingService {
+  @Cached(cacheName = "foo")
+  @Cached(cacheName = "bar")
+  BigData heavyComputation() {
+    // ...
+  }
```
Here's a proposal which is really close from the current Quarkus caching API.

```suggestion
As an example, suppose we have an interceptor that removes all entries from a cache.
The corresponding interceptor binding would be called `@CacheInvalidateAll` and would be declared as `@Repeatable`.
If we wanted to clear two caches at the same time, we would add `@CacheInvalidateAll` twice:

[source,java]
----
@ApplicationScoped
class AllEncompassingService {
  @CacheInvalidateAll(cacheName = "foo")
  @CacheInvalidateAll(cacheName = "bar")
  void clearBothCaches() {
    // ...
  }
```

When this PR is merged, `@CacheInvalidateAll` will even become an interceptor binding (it's a simple annotation right now).

Ladicek replied:
OK, I didn't really want to use _existing_ annotations, but this will do, thanks! :-)

Ladicek commented:
Updated an example per @gwenneg's suggestion. I think this PR is ready now.

gwenneg reviewed (approved):
LGTM, thanks!

Let's wait for @mkouba or @manovotn approval before it is merged though :)

mkouba commented:
I'm not quite sure we want to merge those disabled tests. Ideally, we should fix the problem in a separate PR ;-). @Ladicek WDYT?

Ladicek commented:
Removed disabled tests. I kept them locally in case anyone is interested in the future :-)

gwenneg commented:
Thanks everyone for that fix!
